### ISSUE TITLE
Should CPtr's contents be defined in an internal module or a standard one?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16389]
Currently `CPtr.chpl` is an internal module.  In this issue, I'm proposing that it should be a standard one, using the following argument:

* I think that "internal" modules should be ones that are used to implement things that are "built into" the language and would be described in a complete version of the language spec.  So things like strings, syncs, ranges would all be defined in internal modules.

* Meanwhile, I think of "standard" modules as being those that aren't built into the language, yet are so fundamental or useful that any implementation of Chapel should be able to rely on them being available (and, ultimately, stable).  I think of standard modules as being ones where you'd surf to the standard module documentation on the docs website and review the symbols that that module makes available.

To me, types like `c_ptr()` and `c_void_ptr` are not things that we consider to be an intrinsic features of the Chapel language, but features that are made available in a standard manner to support interoperability and lower-level programming, so it seems obvious that the module should be standard (if others think that `c_ptr` should be part of the language, let me know and I'll spawn off a separate issue about that; the rest of this issue will assume it's not part of the language).

[Also independent of, but related to, this issue, I think CPtr should ultimately be merged into a general `CTypes` module that also supports things like `c_int` (currently defined in `SysCTypes`) and `c_double` (currently defined in `SysBasics`). This is related to issue #16131 ].

A counterargument to this proposal might be:

* internal modules shouldn't rely on standard modules, but some of our built-in types (like `string` and `bytes`) use `c_ptr`, so `CPtr` should remain an internal module

I agree that, whenever possible, internal modules shouldn't rely on standard modules, but I don't think it's always possible / desirable, and I'm not convinced that it warrants putting `CPtr` into the internal modules.  Specifically, if `c_ptr()` is not in the language, I think it would be unfortunate to have the user apply some sort of mental heuristic like "do I think that the Chapel implementers used this module to implement the language?" to determine whether to browse the internal module documentation or the standard module docs to read the `c_ptr` type's documentation.  Meanwhile, I also think it would be counterproductive to support a second non-user-facing `c_ptr` type in order to implement strings without `CPtr`.

We could say "some internal modules will have their documentation placed into the standard module hierarchy in the docs even though they are internal modules", but to me, this essentially says that they're standard modules from the user's perspective, so why not just make them standard modules?

**Comment by lydia-duncan** [comment]
I think I'm in agreement with you on each of the values you're putting forth here.  For me, it's the priority of these values relative to each other that I'm more uncertain about.

I agree that it would be unfortunate for users to have to think about whether a type is used to implement the language when determining where to look for it.  At the same time, I do think we want to minimize the compiler's interaction with the standard libraries.  The standard libraries to me feel like they should be capable of being written by a user without modifying the compiler, but we do a lot of explicit compiler support today for CPtr.  `c_ptr` is explicitly tracked by the compiler, we adjust for it at codegen, we adjust the sizes for certain of the SysCTypes based what we've detected about the architecture on which we are running.  It may not always be reasonable to limit the symbols needed to implement the language to live in the internal modules, but doing so will simplify the implementation.  And it will also increase the likelihood that a user looking for a module to emulate won't accidentally stumble across a strategy that requires more explicit compiler support.

I do think the documentation argument is interesting, but I should hope that if a user doesn't have a good mental model about where to look for a type then they will do a search for the type and not be too bothered by where it ends up.  After all, the distinction between package modules and standard modules is not necessary as cut and dried as we would hope.

I won't lose sleep if CPtr is moved to a standard module.  It just didn't seem like an obvious move to me, so I wanted to put forth why that was, in the hope that we would have enough information to make a permanent decision.

**Comment by dlongnecke-cray** [comment]
> if others think that c_ptr should be part of the language, let me know and I'll spawn off a separate issue about that; the rest of this issue will assume it's not part of the language

I do think that `c_ptr` should be a part of the language. The reason why I think so is that currently, as the language stands today, I think it would be more or less impossible to support something like `c_ptr` without explicit compiler support.

To me, all this compiler magic tells me that `c_ptr` is a language feature supported by our compiler, even if it's not in the language spec. 

I think that the question of whether something should be a standard module or internal module can be answered by how much compiler magic the feature needs to make it work. For the case of something like `list`, it's about 80-90% user code, and the only magic are things like `chpl__autoDestroy` and `_ddata`, which are things that ought to be language features and that we recognize we're deficient about in some regard.

> internal modules shouldn't rely on standard modules, but some of our built-in types (like string and bytes) use c_ptr, so CPtr should remain an internal module

I also agree with this counterargument, but it's probably irrelevant if we agree that `c_ptr` should be part of the language.

**Comment by bradcray** [comment]
I'm not convinced that "requires (or benefits from) compiler support" should be used to determine whether something's part of the internal vs. standard modules.  Ideally, I agree that the less that standard modules (or "any modules" for that matter) rely on compiler support, the better.  But if a module is standard (read: "must be included with any implementation of Chapel"), I don't think it matters deeply if the compiler is involved in its implementation or that there is any obvious reason to believe that a user should be able to write everything in the standard module using only built-in language features.

The following is a slightly weak analogy, but many C compilers have support for generating good error messages for `printf()` but I don't think anyone would argue that `printf()` is part of the C language rather than the standard C library.  (this is weak because AFAIK, printf() doesn't typically rely on the C compiler to implement itself, and the error messages a good C compiler generates seem optional and going above/beyond what the language requires).

Maybe this is a slightly better analogy:  As a user, could you write `fprintf()` in C without using anything in "stdio.h" or "stdlib.h"?  If not, does that mean `fprintf()` (or stdio.h or stdlib.h) should be part of the C language rather than part of its standard library?

**Comment by e-kayrakli** [comment]
TL;DR -- I am leaning mildly against it.

> [Also independent of, but related to, this issue, I think CPtr should ultimately be merged into a general CTypes module that also supports things like c_int (currently defined in SysCTypes) and c_double (currently defined in SysBasics). This is related to issue #16131 ].

I think moving the `c_ptr` to the standard modules as part of a unifying `CTypes` module would make me more comfortable rather than making `CPtr` a standard module.

Other than that, I agree with many opinions (for or against) expressed here. And I also agree that this is kind of a spectrum and where we put the line between standard/internal modules is a subjective. I personally feel that CPtr is a type that supports one of the key features of Chapel and as such should be in the internal modules. We need CPtr to interact with the runtime from the module code, which is pretty intrinsic to how we implement Chapel. However, one can imagine other Chapel implementations where CPtr is not that crucial for the language implementation itself. I might have thought differently if that was the case.

Re `use`/`import` ing standard modules from internals: beyond philosophical arguments, I think we should try to prevent it to avoid maintenance headaches w.r.t module import and initialization order. But whether that outweighs the benefits of moving CPtr type to standard modules in some form is difficult to gauge for me.

**Comment by bradcray** [comment]
> CPtr is a type that supports one of the key features of Chapel and as such should be in the internal modules.

I think stdio.h and stdlib.h support key features of C, but would consider them more equivalent to our standard modules than our internal modules (by my own definition in the OP, granted).

**Comment by dlongnecke-cray** [comment]
> > CPtr is a type that supports one of the key features of Chapel and as such should be in the internal modules.
> 
> I think stdio.h and stdlib.h support key features of C, but would consider them more equivalent to our standard modules than our internal modules (by my own definition in the OP, granted).

I view the `fprintf` analogy as less of a problem compared to something like `CPtr` because we provide facilities (in the form of extern functions) that allow a user to write their own IO plumbing if they really need to. In C, if I wanted to write my own `fprintf` function from the ground up, I can declare the signature for a wrapper around my platform's write system call and then go implement it myself in platform specific assembly if I need to. You might consider this to be reinventing the wheel, and I'd agree with you, but some users might want to reinvent the wheel.

However I'm not sure that there's any way for someone using Chapel to create something that provides behavior like `CPtr` short of modifying the compiler itself, which is part of why I think that even if it's not a language feature (yet), it should stay in the internal modules.

**Comment by bradcray** [comment]
> I can declare the signature for a wrapper around my platform's write system call and then go implement it myself in platform specific assembly if I need to.

To me, you're agreeing with my point which is that C has standard library capabilities that you'd be hard-pressed to write natively in C itself (unless you're considering platform-specific assembly to be part of C by virtue of the `asm` statement which I buy intellectually, but not intuitively—to me  you've still gone outside of "simply C" to achieve your goal).  Basically, I don't think "Are you able to write this feature in the language?" should be the litmus test for whether or not something should be allowed to be in the standard library.

> However I'm not sure that there's any way for someone using Chapel to create something that provides behavior like CPtr short of modifying the compiler itself

When I think through the features in CPtr, it seems like most of them could be implemented with `extern`.  And doing so seems fairly similar to your example of dropping down to platform-specific assembly in C.  Is there some case where this would not work?  (though again, even if there is, I don't think this is the right way to decide whether something should be in internal vs. standard).

Maybe taking a different tack, I think people should be able to write the following statements in Chapel without `use`-ing or `import`-ing any modules because I view them as using core Chapel capabilities:

```chapel
var x: int = 42;
var y: [1..10] real;
writeln(x);
coforall i in 1..x do writeln("hi from task ", i);
```

Whereas I think for code like the following:

```chapel
var x: c_int = 42;
var ptr: c_void_ptr = c_malloc(c_float, 10);
var intPtr: c_ptr(c_int) = c_ptrTo(x);
var arr: c_array(c_int);
if is_c_nil(intPtr) ...
```

some sort of `use` or `import` should be required to gain access to these lower-level capabilities (whether that's `use CTypes;`, `use CInterop;` `use CPtr;`, or whatever we call it).  My argument is almost completely aesthetic, but also about making the default namespace less cluttered than it already is.  I don't view any of these identifiers or capabilities that start with `c_` as being intrinsically about Chapel's core feature set; instead, I view them as very important standard capabilities that we provide with Chapel in order to give programmers access to lower-level programming from within the language.  And that doing so as a standard module is reasonable, even if the compiler is involved in implementing that module.

[that said: in what way do the features of CPtr require compiler support for their implementation?  I'm unable to come up with any significant ones, but may just not be remembering, or aware of, something?]

**Comment by dlongnecke-cray** [comment]
> To me, you're agreeing with my point which is that C has standard library capabilities that you'd be hard-pressed to write natively in C itself (unless you're considering platform-specific assembly to be part of C by virtue of the asm statement which I buy intellectually, but not intuitively—to me you've still gone outside of "simply C" to achieve your goal).

I think I view `CPtr` as more akin to the `extern` function support Chapel provides than the `fprintf` function. My explanation was attempting to show that Chapel provides _a way_ to achieve your goal via `extern` functions.

I agree that in order to implement IO, at some point you have to reach "outside the language" in order to do so. So perhaps my `fprintf` explanation is not the best example.

I think the better way to frame it is to imagine if you removed `writef` from Chapel (and likewise, `fprintf` in C). In both cases you'd survive, because each language provides a mechanism to write such services yourself.

Now imagine if you removed `CPtr` from Chapel (I'll spare us the thought of removing pointers from C...).

> [that said: in what way do the features of CPtr require compiler support for their implementation? I'm unable to come up with any significant ones, but may just not be remembering, or aware of, something?]

```chapel
  pragma "data class"
  pragma "no object"
  pragma "no default functions"
  pragma "no wide class"
  pragma "c_ptr class"
  class c_ptr {
```

As you said, `CPtr` provides the ability for users to interact with C pointers within the language, and we only get that by handling them explicitly in the compiler (I don't want to exaggerate how much support there is, but it's clear from the pragmas that at least _some_ level of compiler support is needed).

I suppose it's true that we could just move all operations we use `CPtr` for into the runtime and then just call out to them, and then go ahead and remove the `CPtr` type as well. I'm not advocating for this, but I think that by imagining its removal, and how the compiler would lose the ability to "speak with" C pointers and recognize their types, it's clear that what we've removed is essentially a language feature, even if we don't recognize it as such.

I just wanted to iron out my thoughts. I know we're definitely dancing in circles at this point!

> My argument is almost completely aesthetic, but also about making the default namespace less cluttered than it already is. I don't view any of these identifiers or capabilities that start with c_ as being intrinsically about Chapel's core feature set; instead, I view them as very important standard capabilities that we provide with Chapel in order to give programmers access to lower-level programming from within the language.

I can agree with this.

The `c_ptr` type itself still feels like a language feature, which is why the idea of moving it to standard modules just so that users have to "opt in" to it by `using` it doesn't sit well with me.

But I better understand the motivation (uncluttering the default namespace).

**Comment by bradcray** [comment]
Another argument:

If we'd intended for the core Chapel language to support pointers, I think that rather than calling the type `c_ptr`, we would've called it `ptr` or `pointer` (i.e., if it's a Chapel pointer, not a C pointer, why put the `c_` prefix on it?  We don't call Chapel's ints `c_int` or `chpl_int`).  However, in designing Chapel, we very intentionally designed it not to have pointers (because they're tricky enough in serial programming, thwart compiler analysis and optimizations, and are arguably strictly worse in parallel/distributed settings).  Instead, classes and references were supported as the closest things a user could get within the language to a pointer-like thing.

Pointers do have a role for more advanced, lower-level programming, which is why support for things like `c_ptr` have been added over time, but this was not intended to be a core part of the language that every user would use; rather, it was intended to be an advanced, "at your own risk", low-level capability for those who need it.  For these reasons, some action (e.g., invoking a `use` or `import`) should be necessary to opt into accessing its features.

**Comment by mppf** [comment]
I definitely think that users should have to write an `import` / `use` to get `CPtr` contents (with the exception that `extern { }` should probably add the use/import for you).

I don't personally care if it's in `modules/standard` or `modules/internal` but I definitely don't think things like `c_ptr` should be visible by default.

**Comment by dlongnecke-cray** [comment]
> Pointers do have a role for more advanced, lower-level programming, which is why support for things like c_ptr have been added over time, but this was not intended to be a core part of the language that every user would use; rather, it was intended to be an advanced, "at your own risk", low-level capability for those who need it. For these reasons, some action (e.g., invoking a use or import) should be necessary to opt into accessing its features.

This resonates with me. Thanks for taking the time to re-iterate your position. I agree that users should have to `use` the `CPtr` module to get access to `c_ptr`.

> I don't personally care if it's in modules/standard or modules/internal but I definitely don't think things like c_ptr should be visible by default.

I still think that `modules/internal` feels more correct to me, but since there's no way to get the import behavior we want if `CPtr` lives there (?), I'm fine with `modules/standard`.

**Comment by lydia-duncan** [comment]
> I still think that modules/internal feels more correct to me, but since there's no way to get the visibility behavior we want if CPtr lives there (?)

I'm pretty sure we can get the visibility behavior you're describing today

**Comment by dlongnecke-cray** [comment]
> > I still think that modules/internal feels more correct to me, but since there's no way to get the visibility behavior we want if CPtr lives there (?)
> 
> I'm pretty sure we can get the visibility behavior you're describing today

Given that, wouldn't another option be to have `CPtr` continue to live in `modules/internal` but have the module require an `import`/`use` in order to use it?

**Comment by bradcray** [comment]
> Given that, wouldn't another option be to have CPtr continue to live in modules/internal but have the module require an import/use in order to use it?

I'm sorry, I don't really understand the value proposition in that proposal.  I think we want to present `CPtr` to users in the documentation as being a standard Chapel module, by which I mean "a library module that's guaranteed to be available with any complete/compliant Chapel release."   I also don't think we want to introduce a new class of module to users in the documentation called "internal modules" which are standard in that they're always available to Chapel, but somehow different from standard modules.  And if we present `CPtr` as a standard module to users, I don't see a benefit in storing it in a directory that's different from where all the other standard modules live (`$CHPL_HOME/modules/standard`).

Put another way:  If someone else decided to do a Chapel implementation, I don't think they'd be at all required to have the same internal modules that we do, or any internal modules at all for that matter: they'd be free to implement every user-facing thing that we happen define in `modules/internal` (e.g., ranges, strings, etc.) in their compiler, the runtime, in Python, or however they wanted.  To that end, the typical Chapel programmer shouldn't need to be aware that `$CHPL_HOME/modules/internal` exists or to ever wonder what's in it.  Whereas standard modules are those which are documented for `use`/`import`, that a user would browse when looking for helpful features that are not part of the language, etc.  That seems to describe the `CPtr` module precisely, so why would we treat it differently?

If your answer is "because it uses pragmas as part of its implementation", do a quick `git grep pragma` in `modules/standard`.  If it's "because other internal modules rely upon it", `git grep` `modules/internal` for `ceil`, `log2`, `writef`, `canResolve`, ...

**Comment by dlongnecke-cray** [comment]
> I'm sorry, I don't really understand the value proposition in that proposal.

Standard modules it is, then!
